### ISSUE TITLE
Feature Request: ECHO and NOECHO

### ISSUE THREAD

**Issue by OPMUSER** [OPM/opm-simulators#1502]
It would be very useful in debugging input decks if the ECHO keyword was implemented to echo the input file as the parser is reading the input deck. The NOECHO keyword would switch echoing of the input deck.

**Comment by atgeirr** [comment]
I am not sure if this could be implemented in a useful way, as Flow reads all of the input before acting on it, rather than following it as an interpreted script. Any ideas @joakim-hove?

**Comment by joakim-hove** [comment]
> I am not sure if this could be implemented in a useful way, as Flow reads all of the input before acting on it, rather than following it as an interpreted script. Any ideas @joakim-hove?

I am not really sure what the expected benefit is here; but I at least agree with @atgeirr that flow works by parsing and internalizing the complete deck before anything happens.

But turning on and off a stream duplication to stdout based on the keywords `ECHO` and `NOECHO` is certainly possible.

**Comment by atgeirr** [comment]
> But turning on and off a stream duplication to stdout based on the keywords ECHO and NOECHO is certainly possible.

I agree, that would be possible. My reluctance is based on the assumption that using ECHO, the user would expect for example, to see WCONHIST keywords interleaved in the output with the actual timesteps, while we can produce it only as a big dump of data before we start simulating.

**Comment by OPMUSER** [comment]
It is to have an audit of the run, so that the engineer knows what was used to generate the output files.  The grid and grid property data can be very voluminous and normally does not change once the model is built, except for perhaps by applying the `MULTIPLY` keywords etc. So normally the grid and grid property data are placed in  separate include files, with the` NOECHO `keyword at the beginning of the file and an `ECHO` at the end. So the engineer knows what file is read in but does not get all the data in the include file listed to the PRT file.

That is just one example, there are many more when the engineer just wishes to see the include file and not all the data in the file, or even in the deck itself.

I don't think we have to be too "clever" here to implement this. We can just read each line and process based on the `ECHO / NOECHO` status. Once the deck has been completely read, then rewind the file and process the deck the way we currently do. That is what the commercial simulator does. 

Will this work?

**Comment by OPMUSER** [comment]
It is to have an audit of the run, so that the engineer knows what was used to generate the output files.  The grid and grid property data can be very voluminous and normally does not change once the model is built, except for perhaps by applying the `MULTIPLY` keywords etc. So normally the grid and grid property data are placed in  separate include files, with the` NOECHO `keyword at the beginning of the file and an `ECHO` at the end. So the engineer knows what file is read in but does not get all the data in the include file listed to the PRT file.

That is just one example, there are many more when the engineer just wishes to see the include file and not all the data in the file, or even in the deck itself.

I don't think we have to be too "clever" here to implement this. We can just read each line and process based on the `ECHO / NOECHO` status. Once the deck has been completely read, then rewind the file and process the deck the way we currently do. That is what the commercial simulator does. 

Will this work?
